# hatpal

**hatpal** is an R package for alternative polyadenylation (APA) identification & analysis using 3’ scRNA-seq (10x etc.) . By clustering the possible polyA sites, hatpal can find high-confidence APA clusters without GTF file, and create a count matrix for downstream analysis after being annotated.



## Before the start

### Dependencies

Softwares: samtools and R

R packages: data.table, GenomicAlignments, kpeaks and Ckmeans.1d.dp

```R
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
if (!require("data.table")) install.packages("data.table")
if (!require("kpeaks")) install.packages("kpeaks")
if (!require("Ckmeans.1d.dp")) install.packages("Ckmeans.1d.dp")
if (!require("GenomicAlignments")) 
    BiocManager::install("GenomicAlignments")
```

### Installation

hatpal can be installed from github through devtools:

```R
library(devtools)
devtools::install_github('HHengUG/hatpal')
```

hatpal can also be installed as followed in linux:

```shell
R CMD INSTALL hatpal_0.9.1.tar.gz
```

and in windows:

```R
install.packages("hatpal_0.9.1.zip", repos = NULL, type = "win.binary")
```

### Input preparation

- An **indexed BAM** file generated from 3’ scRNA-seq for preprocessing

- A **GTF** file if you need annotation and count matrix for downstream analysis in Seurat

 To find the mapped soft-clipped reads,  run the preprocessing of BAM file on Linux via **samtools**.

```shell
samtools view -F 16 input.sorted.bam | awk '{if ($6~/([0-9][0-9]|[3-9])S$/ ) print }' > run/run08/p_3S_test.sam &

samtools view -f 16 input.sorted.bam | awk '{if ($6~/^([0-9][0-9]|[3-9])S/ ) print }' > run/run08/n_3S_test.sam &

samtools view -H input.bam > run/run08/SAM_header &

cat run/run08/SAM_header run/run08/p_3S_test.sam run/run08/n_3S_test.sam | samtools view -b > run/run08/3S_input.bam &

samtools sort 10x/run/run08/3S_test.bam > 10x/run/run08/3S_input.sorted.bam &

samtools index -b 10x/run/run08/3S_input.sorted.bam &
```

The output is an indexed BAM file containing soft-clipped reads needed in hatpal.



## Run hatpal

hatpal includes five steps and functions.

### ExtractChrinfo

*ExtractChrinfo()* extracts the chromosomes information from the indexed BAM file, including names and length.

The input should be a BAM file with a .bai index in the same directory.

```R
chr_info <- ExtractChrinfo("input/input.sorted.bam")
```

The output is a data.table object in R.

### WriteEndlist

*WriteEndlist()* generates two-dimensional coordinate files of each strand for clustering.

Two BAM files and an output directory are needed, and .bai index in the same directory is also required. The next parameter is the output directory, and the last parameter is a data.table containing chromosomes information extracted by *ExtractChrinfo()*.

Use the BAM file containing soft-clipped reads twice if you want high-confidence APA clusters :

```R
WriteEndlist("input/3S_input.sorted.bam", "input/3S_input.sorted.bam", "output/", ExtractChrinfo("input/3S_test.sorted.bam"))
```

Or if you want more features, use the original indexed BAM file as the first parameter, and the soft-clipped indexed BAM as the second parameter.

```R
WriteEndlist("input/input.sorted.bam", "input/3S_test.sorted.bam", "output/", ExtractChrinfo("input/input.sorted.bam"))
```

The outputs are four files in two directories (negative_strand/ and positive_strand/), and an extra file containing the list of cell barcodes named "output/_CBlist.txt".

### ClusterAPA

*ClusterAPA()* clusters the data in two two-dimensional coordinate files (RC and SA) to find the APA clusters.

The input files are in two directories (negative_strand/ and positive_strand/) generated by *WriteEndlist()*. The last parameter is the output directory. Every strand has two two-dimensional coordinate files,  so you should run twice for each strand. 

```R
ClusterAPA("output/negative_strand/sa.prebed", "output/negative_strand/sa.prebed", "output/negative_strand/")
ClusterAPA("output/positive_strand/sa.prebed", "output/positive_strand/sa.prebed", "output/positive_strand/")
```

The outputs are two files containing APA clusters (APA_clusters.out).

### AnnoAPAc

*AnnoAPAc()* annotates the APA clusters from a GTF file.

The first two inputs are  one APA cluster file and one GTF file. The third parameter determines the strand. If chromosome names of the GTF file do not have "chr" in them, the fourth parameter (add_chr = TRUE) will add string "chr" before them. You should also run *AnnoAPAc()* twice for each strand.

```R
AnnoAPAc("output/positive_strand/APA_clusters.out", "ref/hg_19.gtf", "+")
AnnoAPAc("output/negative_strand/APA_clusters.out", "ref/hg_19.gtf", "-")
```

When the chromosome names of the GTF do not have "chr" in it (i.e., 1 2 3), but that of the APA_clusters.out or chr_info have (i.e., chr1 chr2 chr3), run:

```R
AnnoAPAc("output/positive_strand/APA_clusters.out", "ref/hg_19.gtf", "+", add_chr = TRUE)
AnnoAPAc("output/negative_strand/APA_clusters.out", "ref/hg_19.gtf", "-", add_chr = TRUE)
```

The output is a file where each annotation of clusters is described by eight columns. The first common three columns are chr, start and end, and the fourth and fifth columns are annotated gene and type, if there is no annotation for the clusters, then the last three columns will be the gene, type and distance of the closest annotation on the 5' end.

### CountCB

*CountCB()* creates a APA_cluster-Cell_Barcode count matrix which can be used in Seurat analysis.

The first two parameters are annotated APA cluster files of positive strand and negative strand. The third parameter is a BAM file with a .bai index in the same directory. The fourth parameter is a CB list, you can use white list of cell barcode from cellranger, or  use the "output/_CBlist.txt" generated by *WriteEndlist()*. The fifth parameter is a data.table containing chromosomes information extracted by *ExtractChrinfo(),* and the last parameter is the output directory.

```R
CountCB("output/positive_strand/APA_clusters.out.anno", 
        "output/negative_strand/APA_clusters.out.anno", 
        "input/test.sorted.bam","output/_CBlist.txt", 
        ExtractChrinfo("input/input.sorted.bam"), "output/")
```

The outputs are three files (genes.tsv, barcodes.tsv and martrix.mtx) in a directory named "matrix/" under "output/". 

This can be easily loaded into Seurat:

```R
library(Seurat)
pbmc.data  <- Read10X("output/matrix/")
```



## Sample 

The [3k PBMC dataset](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.0.0/pbmc3k) from 10x Genomics can be used for testing.



## News

2020/11/5: New releases

- hatpal 0.9.1

